[compiled autograd] have context manager turn on dynamo config

xmfan opened the pull request:
We need a way to unify CA checks across ctx manager or torch.compile APIs. Setting this config is actually a no-op, but unifies the state

One edge case is user using torch.compile on regions that have .backward() but don't want CA to execute on them and another thread is within the context manager. The general solution to the problem would be to first make the state thread local, then have FSDP create threads that carryover the compiled autograd state

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138187
* __->__ #138241
* #138113



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec @yf225

xmfan committed:
[compiled autograd] have context manager turn on dynamo config

[ghstack-poisoned]

yf225 commented on torch/_dynamo/compiled_autograd.py:
```diff
@@ -534,12 +534,15 @@ def enable(compiler_fn):
         if not prior:
             compiled_autograd_enabled = False
+        torch._dynamo.config.compiled_autograd = prior_config
         torch._C._dynamo.compiled_autograd.set_autograd_compiler(prior)
```
maybe we can rebase on top of https://github.com/pytorch/pytorch/pull/138113 to avoid potential merge conflict?

xmfan committed:
Update on "[compiled autograd] have context manager turn on dynamo config"

We need a way to unify CA checks across ctx manager or torch.compile APIs. Setting this config is actually a no-op, but unifies the state

One edge case is user using torch.compile on regions that have .backward() but don't want CA to execute on them and another thread is within the context manager. The general solution to the problem would be to first make the state thread local, then have FSDP create threads that carryover the compiled autograd state




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec yf225

[ghstack-poisoned]

yf225 committed:
Update on "[compiled autograd] have context manager turn on dynamo config"

We need a way to unify CA checks across ctx manager or torch.compile APIs. Setting this config is actually a no-op, but unifies the state

One edge case is user using torch.compile on regions that have .backward() but don't want CA to execute on them and another thread is within the context manager. The general solution to the problem would be to first make the state thread local, then have FSDP create threads that carryover the compiled autograd state




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec yf225

[ghstack-poisoned]

yf225 committed:
Update on "[compiled autograd] have context manager turn on dynamo config"

We need a way to unify CA checks across ctx manager or torch.compile APIs. Setting this config is actually a no-op, but unifies the state

One edge case is user using torch.compile on regions that have .backward() but don't want CA to execute on them and another thread is within the context manager. The general solution to the problem would be to first make the state thread local, then have FSDP create threads that carryover the compiled autograd state




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec yf225

[ghstack-poisoned]

yf225 commented:
hmm seems to be failing in `TestCompiledAutograd.test_reentrant_checkpointing`: https://productionresultssa5.blob.core.windows.net/actions-results/af427d57-7b9e-495e-8be9-d1c9519cf26c/workflow-job-run-7ce09cf7-794a-5c4b-ef14-bd5267058927/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-10-18T00%3A20%3A10Z&sig=ZVm%2BJgRT10qCSz4BbEFFg7xDBQcAtk1YxE7VCjZpMMU%3D&ske=2024-10-18T10%3A16%3A51Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2024-10-17T22%3A16%3A51Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2024-08-04&sp=r&spr=https&sr=b&st=2024-10-18T00%3A10%3A05Z&sv=2024-08-04

xmfan commented:
nvm this doesnt work for reentrants
